### DISCUSSION TITLE
jakarta.persistence

### DISCUSSION THREAD

**Discussion by gavinking** [hibernate/hibernate-orm#3773] [Design Proposals]
For Hibernate 6 we need to decide how we're going to support the `jakarta.persistence` APIs.

Currently `Session` directly extends `javax.persistence.EntityManager`. This means that there's currently no way to use Hibernate without a dependence on `javax.persistence`.

But _eventually_ we'll need to also support `jakarta.persistence`. We need to figure out how to do that, and I would rather make sure our strategy is clear before releasing H6.

**Comment by Sanne** [comment]
We'll want to publish both `jakarta.` and `javax.` compatible artifacts for a while [1].

By automating this at build-tool level, it would be easier for end users as they'll be able to switch API at a time which suits them, without having to switch ORM implementation at the same time (which in an ideal world is flawless but in practice could be problematic).

I suggest we start prototyping this in the `master` (5.5) branch: this branch is unreleased precisely as the "theme" of this minor was meant to be Jakarta. So we could have a couple Beta's in the 5.5 series to polish this and assuming most code changes will be in the build scripts, it should be straight forward to apply on the 6 branch.


1 - and I'd rather have us prepared for a very fluid definition of "while" - as you mentioned in another context, and I completely agree, we will have people wanting to migrate not only this summer but also for some years to come.

**Comment by gavinking** [comment]
> By automating this at build-tool level, 

Oh gawd I hate that idea.

It only works for as long as there is _no difference_ between `java.persistence` and `jakarta.persistence`: that is, it just kicks the can down the road.

I think we should support both APIs at the same time, via wrappers.

**Reply by gavinking** [reply]
What I'm mean is that it's a solution that seems easier in the short term, but I fear it will be harder to maintain over the longer term.

**Reply by sebersole** [reply]
Wrapping Session, which is how this was done originally, caused major headaches.  Seems easier, but not

**Reply by Sanne** [reply]
That was my preference as well - this isn't a decision we took lightly ;) 

Happy to re-discuss this but include @n1hility , as he's coordinating things

**Reply by Sanne** [reply]
BTW the (alleged?) problem with having both wrappers is that IDE's will auto-suggest e.g. annotations to autocomplete the package from both APIs, as they will have both on classpath.

I have hoped to challenge this as a minor problem as

A) that's not true for Gradle projects, or any other build tool except Maven

B) It might be doable to have IDE's deal with this in better ways

**Reply by gavinking** [reply]
> Wrapping Session, which is how this was done originally, caused major headaches. Seems easier, but not

I can see why it wasn't easier when we only had *one* API to support. Now we have two, which changes the tradeoffs significantly.

**Reply by gavinking** [reply]
> BTW the (alleged?) problem with having both wrappers is that IDE's will auto-suggest e.g. annotations to autocomplete the package from both APIs, as they will have both on classpath.

I insist that this is solvable. I could be wrong.

**Reply by sebersole** [reply]
> > Wrapping Session, which is how this was done originally, caused major headaches. Seems easier, but not
> 
> I can see why it wasn't easier when we only had _one_ API to support. Now we have two, which changes the tradeoffs significantly.

No, it changes the tradeoffs significantly *temporarily*.  Later when we phase out the old contracts we have the same thing again

**Reply by gavinking** [reply]
But "temporary" probably means _for the next 5-10 years_ in this case.

**Reply by beikov** [reply]
Crazy idea here, we could actually implement both, the javax.persistence APIs and the jakarta.persistence APIs by generating bridge methods in the bytecode. In Java it is illegal to do this (same method with different return type), but on the bytecode level this is fine. So we could code against e.g. jakarta.persistence APIs so we could support the new features and our build will add `implements` for all corresponding javax.persistence APIs and generate bridge methods that do casting to delegate to the actual implementation. I think this would "penalize" calls to java.persistence APIs due to the fact that it has to go through the bridge method and do casting, but when the JIT kicks in, these monomorphic call sites should go away.

Not sure if the Eclipse Transformer project can do that, but I guess Hibernate is not the only project that would like to be able to support both APIs, so maybe we can work on this as part of the Eclipse Transformer project if needed. The only "problem" is that we would still need to publish two artifacts because of the build dependencies (pom.xml).

**Reply by Sanne** [reply]
right if we could have Hibernate being *binary compatible* with either/or that would provide a superb upgrade path for users.

But we'd still have the drawback of
1) requiring both APIs on classpath
2) need to enforce some level of consistency, e.g. disallow the user to use both javax.persistence and jakarta.persistence on the entities as that could lead to conflicting information (`javax.persistencer.GeneratedValue(strategy=SEQUENCE)` && `jakarta.persistencer.GeneratedValue(strategy=IDENTITY)` and many other flavours of no-longer-unique annotations)

**Reply by Sanne** [reply]
perhaps actually Java Modules could help with the classpath pollution aspect?  If we declared both the Jakarta Persistence API and the JPA API as private module dependencies, the users would have to explicitly opt-in into *either* of them - and the IDE would do the right thing?

Sounds promising but the drawback is clearly that we'd need to be fully ready to be modular as well.

**Reply by sebersole** [reply]
I'm against wrappers.  Yes they sound like the easy, silver-bullet solution.  But that approach was problematic in the past.

**Reply by Sanne** [reply]
N.B. another argument in favour of being able to support either API with the same artifact is that complex applications typically are composed of multiple modules, such as a main deployable and a set of additional jars which contain mapped entities.

Being able to support either would greatly help people by not having to first rebuild all of their dependent modules.

For example it would be totally feasible for us to load some entities annotated via flavour B and others via flavour A, as long as there's consistency within each individual class (as I hinted in previous comment). Large applications could then migrate in small steps.

**Reply by gavinking** [reply]
I'm not attached to any one particular approach. My concern is that this "solution" isn't a real solution that is going to work out in the longer term once Jakarta starts adding new APIs, and then we're going to be back revisiting this very same discussion in like 2 years time.

**Reply by gavinking** [reply]
(But by then the design of H6 will be set in stone.)

**Reply by Sanne** [reply]
> I'm not attached to any one particular approach. My concern is that this "solution" isn't a real solution that is going to work out in the longer term once Jakarta starts adding new APIs, and then we're going to be back revisiting this very same discussion in like 2 years time.

True, but such an approach would allow us to support both versions at least for the time they are aligned - and it should be quite likely that Jakarta improvements will be additive to the JPA (since JPA is not going to change at all) so at that point we should be able to add new things to one "side" only. Honestly if we happen to accidentally support a new feature on JPA as well, that doesn't seem like a big problem.

**Reply by n1hility** [reply]
> right if we could have Hibernate being _binary compatible_ with either/or that would provide a superb upgrade path for users.
> 
> But we'd still have the drawback of
> 
> 1. requiring both APIs on classpath

Please don't do that :) It would make the experience far worse than the status quo.

**Reply by n1hility** [reply]
> BTW the (alleged?) problem with having both wrappers is that IDE's will auto-suggest e.g. annotations to autocomplete the package from both APIs, as they will have both on classpath.

Doesn't this depend on how it's structured?  For example, if you did composition wrappers then doesn't that mean you can have entitymanager-jakarta and entitymanager-javax and then you just import the right GAV to get the right dependency set?

**Reply by sebersole** [reply]
> > BTW the (alleged?) problem with having both wrappers is that IDE's will auto-suggest e.g. annotations to autocomplete the package from both APIs, as they will have both on classpath.
> 
> Doesn't this depend on how it's structured? For example, if you did composition wrappers then doesn't that mean you can have entitymanager-jakarta and entitymanager-javax and then you just import the right GAV to get the right dependency set?

You can, and in fact this is how we initially had so that we could limit the dependency on JPA.  There are difficulties in the wrapper approach though and especially when splitting them out into separate modules.

**Reply by Sanne** [reply]
> > BTW the (alleged?) problem with having both wrappers is that IDE's will auto-suggest e.g. annotations to autocomplete the package from both APIs, as they will have both on classpath.
> 
> Doesn't this depend on how it's structured? For example, if you did composition wrappers then doesn't that mean you can have entitymanager-jakarta and entitymanager-javax and then you just import the right GAV to get the right dependency set?

Sure that could work. Except, like @sebersole pointed out, it would be a huge mess for us to extract such components. Integration is really tied ATM, and for good reasons.

What do you think of using Java modules? We could have our library link to both and yet not expose them to user's classpath.

**Reply by n1hility** [reply]
It's a clever idea. I think the issue is then you are forcing a conversion of everything to JPMS, which is arguably more invasive than a simple source conversion switch to jakarta. There is probably still some funkiness in the IDE. If you do a class search for example (say command N on IDEA, you probably still pick up the dups, and possibly set breakpoints in the wrong places etc). There is the added issue that you are forcing the download of duplicate API artifacts that have to be loaded (when running on a JVM). Maybe this amounts to a trivial overhead, but at the end of the day you are forcing the user to pull in elements they don't actually want, when it never makes sense to use both. The approach also seems a little bit more susceptible to issues compared to the other options. For example, if you have a slight ABI difference:

-  Build time transformation approach -  sortable with a bit of patching logic.
-  Wrapper approach - solvable with wrapper logic
-  Dynamic transformation - same as build-time, some sort of rules
-  Hard fork - N/A 

Unless I am missing something it seems to offer no benefit over the other options, but a lot of drawbacks.

**Reply by n1hility** [reply]
> > > BTW the (alleged?) problem with having both wrappers is that IDE's will auto-suggest e.g. annotations to autocomplete the package from both APIs, as they will have both on classpath.
> > 
> > 
> > Doesn't this depend on how it's structured? For example, if you did composition wrappers then doesn't that mean you can have entitymanager-jakarta and entitymanager-javax and then you just import the right GAV to get the right dependency set?
> 
> Sure that could work. Except, like @sebersole pointed out, it would be a huge mess for us to extract such components. Integration is really tied ATM, and for good reasons.

(Sorry I thought I must have lost my reply on this one). 

Yeah I can appreciate wanting to avoid the added complexity and the difficulty in ensuring hotspot callsites remain monomorphic.  Whats particular tricky about investing in any complex solution is the incentive for developers to move to Jakarta is pretty low atm, since there is no new features to move to. This is why I think the options which negatively impact usability aren't a great solution either since. from the users perspective, it's just all pain, and they don't get much out of it. 

(Although granted I realize we aren't necessarily talking short term with some of these. I am not actually trying to advocate for a particular implementation approach, more just discussing pros/cons. You guys know best there on how you want to do it. I do care more about the contact from a Quarkus consumption perspective)

**Reply by KrishnaST** [reply]
Just an opinion. 
Its not only about jakarta.persistence. If you guys want to support jakarta.persistence I suggest you completely dump javax.*. for example in one of my projects I am using Jersey 3.0.1 which has dependency on jakarta.xml.bind, in the same project I am using hibernate 5.4.30.Final which has dependency on javax.xml.bind. I have to add two sets of JAXB implementations in my project. same goes for other dependencies which are now fall under jakarta such as jakarta.activation.

the best scenario is you dump javax.* in future final release and directly support jakarta.persistence without wrappers. those who want to stick to javax.* can use hibernate 5.x version.

**Comment by beikov** [comment]
With Hibernate 5.5.0.Final we now have _-jakarta_ artifacts that are created through build-time transformations. Also see https://in.relation.to/2021/06/04/hibernate-is-jakarta-jpa-2/
